src/modules/bulletproofs/bulletproofs_util.h

```diff
@@ -59,4 +59,55 @@ static void secp256k1_bulletproofs_commit_initial_data(
     secp256k1_sha256_finalize(&sha256, commit);
 }
 
+/* Iterator which produces each l,r pair in succession with a minimum
+ * of scalar operations */
+typedef struct {
+    const unsigned char *nonce;
+    const secp256k1_scalar *y;
+    const secp256k1_scalar *z;
+    secp256k1_scalar yn;
+    secp256k1_scalar z22n;
+    uint64_t val_less_min;
+    size_t count;
+} secp256k1_bulletproofs_bulletproofs_lrgen;
+
+static void secp256k1_bulletproofs_lrgen_init(secp256k1_bulletproofs_bulletproofs_lrgen *generator, const unsigned char *nonce, const secp256k1_scalar *y, const secp256k1_scalar *z, const uint64_t val_less_min) {
+    generator->nonce = nonce;
+    generator->y = y;
+    generator->z = z;
+    secp256k1_scalar_set_int(&generator->yn, 1);
+    generator->val_less_min = val_less_min;
+    generator->count = 0;
+}
+
+static void secp256k1_lr_generate(secp256k1_bulletproofs_bulletproofs_lrgen *generator, secp256k1_scalar *lout, secp256k1_scalar *rout, const secp256k1_scalar *x) {
```

Comment:
`y` and `z` are constants. `x` is a variable. By treating `x` separately we retain the ability to make copies of the lrgen generator rather than re-initializing it. (Indeed, we do this in Step 3 on line 380.)